app/src/main/java/org/thoughtcrime/securesms/keyvalue/KeyValueStore.java

```diff
@@ -0,0 +1,217 @@
+ * or the system killing us unexpectedly (i.e. a force-stop).
+ */
+public final class KeyValueStore implements KeyValueReader {
+
+  private static final String TAG = Log.tag(KeyValueStore.class);
+
+  private final ExecutorService  executor;
+  private final KeyValueDatabase database;
+
+  private KeyValueDataSet dataSet;
+
+  public KeyValueStore(@NonNull Application application) {
+    this.executor = SignalExecutors.newCachedSingleThreadExecutor("signal-KeyValueStore");
+    this.database = KeyValueDatabase.getInstance(application);
+  }
+
+  @AnyThread
+  @Override
+  public synchronized byte[] getBlob(@NonNull String key, byte[] defaultValue) {
+    initializeIfNecessary();
+    return dataSet.getBlob(key, defaultValue);
+  }
+
+  @AnyThread
+  @Override
+  public synchronized boolean getBoolean(@NonNull String key, boolean defaultValue) {
+    initializeIfNecessary();
+    return dataSet.getBoolean(key, defaultValue);
+  }
+
+  @AnyThread
+  @Override
+  public synchronized float getFloat(@NonNull String key, float defaultValue) {
+    initializeIfNecessary();
+    return dataSet.getFloat(key, defaultValue);
+  }
+
+  @AnyThread
+  @Override
+  public synchronized int getInteger(@NonNull String key, int defaultValue) {
+    initializeIfNecessary();
+    return dataSet.getInteger(key, defaultValue);
+  }
+
+  @AnyThread
+  @Override
+  public synchronized long getLong(@NonNull String key, long defaultValue) {
+    initializeIfNecessary();
+    return dataSet.getLong(key, defaultValue);
+  }
+
+  @AnyThread
+  @Override
+  public synchronized String getString(@NonNull String key, String defaultValue) {
+    initializeIfNecessary();
+    return dataSet.getString(key, defaultValue);
+  }
+
+  @AnyThread
+  @Override
+  public synchronized boolean containsKey(@NonNull String key) {
+    initializeIfNecessary();
+    return dataSet.containsKey(key);
+  }
+
+  /**
+   * @return A writer that allows writing and removing multiple entries in a single atomic
+   *         transaction.
+   */
+  @AnyThread
+  @NonNull Writer beginWrite() {
+    return new Writer();
+  }
+
+  /**
+   * @return A reader that lets you read from an immutable snapshot of the store, ensuring that data
+   *         is consistent between reads. If you're only reading a single value, it is more
+   *         efficient to use the various get* methods instead.
+   */
+  @AnyThread
+  synchronized @NonNull KeyValueReader beginRead() {
+    initializeIfNecessary();
+
+    KeyValueDataSet copy = new KeyValueDataSet();
+    copy.putAll(dataSet);
+
+    return copy;
+  }
+
+  /**
+   * Ensures that any pending writes (such as those made via {@link Writer#apply()}) are finished.
+   */
+  @AnyThread
+  synchronized void blockUntilAllWritesFinished() {
+    CountDownLatch latch = new CountDownLatch(1);
+
+    executor.execute(latch::countDown);
+
+    try {
+      latch.await();
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)